dex/ws/wslink.go

```diff
@@ -100,15 +101,19 @@ func (c *WSLink) Send(msg *msgjson.Message) error {
 		return err
 	}
 
-	// If outChan is blocking (outBufferSize pending messages), many messages
-	// are being sent and not quickly enough. If the connection is actually
-	// down, c.quit will be closed. TODO: Add a default case with error return
-	// or just block until outHandler receives?
-	select {
-	case c.outChan <- b:
-	case <-c.quit:
-		return ErrPeerDisconnected
-	}
+	c.sendWG.Add(1)
+	c.outChan <- b
+	go func() {
+		c.writeMtx.Lock()
+		b := <-c.outChan
+		c.conn.SetWriteDeadline(time.Now().Add(writeWait))
+		err := c.conn.WriteMessage(websocket.TextMessage, b)
+		c.writeMtx.Unlock()
+		c.sendWG.Done()
+		if err != nil {
+			c.Disconnect()
+		}
+	}()
```

Comment:
This pattern is quite different than before, when there was a persistent goroutine to handle pings and outgoing messages. Doing it like this with the `sendWg sync.WaitGroup` allows us to have a "graceful" shutdown, and not be racy with any messages sent right before `Disconnect` is called. The `outChan` is still used to sequence the sends.